crates/apollo-compiler/src/validation/directive.rs

```diff
@@ -7,6 +7,130 @@ use crate::{
     ValidationDatabase,
 };
 
+/// Track used names in a recursive function.
+///
+/// Pass the result of `stack.push(name)` to recursive calls. Use `stack.contains(name)` to check
+/// if the name was used somewhere up the call stack.
+struct RecursionStack<'a>(&'a mut Vec<String>);
+impl RecursionStack<'_> {
+    fn push(&mut self, name: String) -> RecursionStack<'_> {
+        self.0.push(name);
+        RecursionStack(self.0)
+    }
+    fn contains(&self, name: &str) -> bool {
+        self.0.iter().any(|seen| seen == name)
+    }
+    fn first(&self) -> Option<&str> {
+        self.0.get(0).map(|s| s.as_str())
+    }
+}
+impl Drop for RecursionStack<'_> {
+    fn drop(&mut self) {
+        self.0.pop();
+    }
+}
+
+/// This struct just groups functions that are used to find self-referential directives.
+/// The way to use it is to call `FindRecursiveDirective::check`.
+struct FindRecursiveDirective<'a> {
+    db: &'a dyn ValidationDatabase,
+}
+impl FindRecursiveDirective<'_> {
```

Comment:
wondering if we can use this sort of thing for fragments forming cycles (self referential fragments, if you will) 🤔